modules/core/src/test/java/org/apache/ignite/internal/processors/query/ScanQueriesTopologyMappingTest.java

```diff
@@ -0,0 +1,274 @@
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+import static org.apache.ignite.internal.IgniteNodeAttributes.ATTR_IGNITE_INSTANCE_NAME;
+
+/** */
+public class ScanQueriesTopologyMappingTest extends GridCommonAbstractTest {
+    /** */
+    private boolean client;
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+
+        super.afterTest();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        return super.getConfiguration(igniteInstanceName)
+            .setCommunicationSpi(new TestRecordingCommunicationSpi())
+            .setClientMode(client);
+    }
+
+    /** */
+    @Test
+    public void testPartitionedQueryWithRebalance() throws Exception {
+        checkQueryWithRebalance(CacheMode.PARTITIONED);
+    }
+
+    /** */
+    @Test
+    public void testReplicatedQueryWithRebalance() throws Exception {
+        checkQueryWithRebalance(CacheMode.REPLICATED);
+    }
+
+    /** */
+    @Test
+    public void testPartitionedQueryWithNodeFilter() throws Exception {
+        checkQueryWithNodeFilter(CacheMode.PARTITIONED);
+    }
+
+    /** */
+    @Test
+    public void testReplicatedQueryWithNodeFilter() throws Exception {
+        checkQueryWithNodeFilter(CacheMode.REPLICATED);
+    }
+
+    /** */
+    @Test
+    public void testLocalCacheQueryMapping() throws Exception {
+        IgniteEx ign0 = startGrid(0);
+
+        IgniteCache<Object, Object> cache = ign0.createCache(new CacheConfiguration<>(GridAbstractTest.DEFAULT_CACHE_NAME)
+            .setCacheMode(CacheMode.LOCAL));
+
+        cache.put(1, 2);
+
+        startGrid(1);
+
+        ScanQuery<Object, Object> qry = new ScanQuery<>();
+
+        {
+            List<Cache.Entry<Object, Object>> res0 = grid(0).cache(GridAbstractTest.DEFAULT_CACHE_NAME).query(qry).getAll();
+
+            assertEquals(1, res0.size());
+            assertEquals(1, res0.get(0).getKey());
+            assertEquals(2, res0.get(0).getValue());
+        }
+
+        {
+            List<Cache.Entry<Object, Object>> res1 = grid(1).cache(GridAbstractTest.DEFAULT_CACHE_NAME).query(qry).getAll();
+
+            assertTrue(res1.isEmpty());
+        }
+    }
+
+    /** */
+    private void checkQueryWithRebalance(CacheMode cacheMode) throws Exception {
+        IgniteEx ign0 = startGrid(0);
+
+        IgniteCache<Object, Object> cache0 = ign0.createCache(new CacheConfiguration<>(GridAbstractTest.DEFAULT_CACHE_NAME)
+            .setCacheMode(cacheMode));
+
+        cache0.put(1, 2);
+
+        blockRebalanceSupplyMessages(ign0, DEFAULT_CACHE_NAME, getTestIgniteInstanceName(1));
+
+        startGrid(1);
+
+        client = true;
+
+        startGrid(10);
+
+        int part = ign0.affinity(DEFAULT_CACHE_NAME).partition(1);
+
+        for (int i = 0; i < 100; i++) {
```

Comment:
To be honest I don't quite understand why we should 100 attempts to check conditions. ^)